pkg/agent/multicluster/mc_route_controller.go

```diff
@@ -214,40 +252,237 @@ func (c *MCRouteController) Run(stopCh <-chan struct{}) {
+		go wait.Until(c.ciImpWorker, time.Second, stopCh)
 	}
 	<-stopCh
 }
 
-// worker is a long-running function that will continually call the processNextWorkItem
-// function in order to read and process a message on the workqueue.
-func (c *MCRouteController) worker() {
-	for c.processNextWorkItem() {
+// gatewayWorker is a long-running function that will continually call the processNextGatewayItem
+// function in order to read and process a message on the gatewayQueue.
+func (c *MCRouteController) gatewayWorker() {
+	for c.processNextGatewayItem() {
 	}
 }
 
-func (c *MCRouteController) processNextWorkItem() bool {
-	obj, quit := c.queue.Get()
+// ciImpWorker is a long-running function that will continually call the processNextClusterInfoImportItem
+// function in order to read and process a message on the ciImpQueue.
+func (c *MCRouteController) ciImpWorker() {
+	for c.processNextClusterInfoImportItem() {
+	}
+}
+
+func (c *MCRouteController) processNextClusterInfoImportItem() bool {
+	obj, quit := c.ciImpQueue.Get()
 	if quit {
 		return false
 	}
-	defer c.queue.Done(obj)
+	defer c.ciImpQueue.Done(obj)
 
-	if k, ok := obj.(string); !ok {
-		c.queue.Forget(obj)
+	ciImp, ok := obj.(string)
+	if !ok {
+		c.ciImpQueue.Forget(obj)
 		klog.InfoS("Expected string in work queue but got", "object", obj)
 		return true
-	} else if err := c.syncMCFlows(); err == nil {
-		c.queue.Forget(k)
+	}
+	if c.wireGuardConfig != nil {
+		if err := c.updateWireGuardPeers(ciImp); err != nil {
+			c.ciImpQueue.AddRateLimited(ciImp)
+			klog.ErrorS(err, "Error updating WireGuard peers, requeuing", "key", ciImp)
+		}
+	}
+
+	if err := c.syncMCFlows(); err == nil {
+		c.ciImpQueue.Forget(obj)
 	} else {
 		// Put the item back on the workqueue to handle any transient errors.
-		c.queue.AddRateLimited(k)
-		klog.ErrorS(err, "Error syncing key, requeuing", "key", k)
+		c.ciImpQueue.AddRateLimited(ciImp)
+		klog.ErrorS(err, "Error syncing key, requeuing", "key", ciImp)
 	}
 	return true
 }
 
+func (c *MCRouteController) processNextGatewayItem() bool {
+	obj, quit := c.gatewayQueue.Get()
+	if quit {
+		return false
+	}
+	defer c.gatewayQueue.Done(obj)
+
+	gateway, ok := obj.(string)
+	if !ok {
+		c.gatewayQueue.Forget(obj)
+		klog.InfoS("Expected string in work queue but got", "object", obj)
+		return true
+	}
+
+	if c.wireGuardConfig != nil {
+		if err := c.reconcileWireGuard(); err != nil {
+			klog.ErrorS(err, "Error when reconcile WireGuard, requeuing", "key", gateway)
+			c.gatewayQueue.AddRateLimited(gateway)
+		}
+	}
+	if err := c.syncMCFlows(); err == nil {
+		c.gatewayQueue.Forget(obj)
+	} else {
+		// Put the item back on the workqueue to handle any transient errors.
+		c.gatewayQueue.AddRateLimited(gateway)
+		klog.ErrorS(err, "Error syncing key, requeuing", "key", gateway)
+	}
+	return true
+}
+
+func (c *MCRouteController) reconcileWireGuard() error {
+	c.wireGuardMutex.Lock()
+	defer c.wireGuardMutex.Unlock()
```

Comment:
This scope of the lock is too wide. Please revisit how to shrink the scope of the locks like `wireGuardMutex`, `mutex`, `ciMutex`. Maybe rename them with more meaningful names. Another question is that is there any possible to use RWMutex to replace any of `wireGuardMutex`, `mutex`, `ciMutex`?

Reply:
Renamed now. The mutexes work for WireGuard, when adding peers and setup interfaces. If we do not use it in this method, maybe we cannot guarantee the process works well